File: tests/DesktopGrass.WPF.Tests/SimTests/FlowerTests.cs
// FlowerTests.cs - §5 flower stream + non-interference with the main stream.

using System;
using System.Linq;
using DesktopGrass.WPF;
using Xunit;

namespace DesktopGrass.WPF.Tests.SimTests;

public class FlowerTests
{
    private const double Monitor1920 = 1920.0;

    [Fact]
    public void FlowerStreamIsDeterministic()
    {
        var a = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, 1.0);
        var b = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, 1.0);

        Assert.Equal(a.Length, b.Length);
        for (int i = 0; i < a.Length; i++)
        {
            Assert.Equal(a[i].IsFlower, b[i].IsFlower);
            Assert.Equal(a[i].FlowerHeadColorIdx, b[i].FlowerHeadColorIdx);
            Assert.Equal(a[i].FlowerHeadRadius, b[i].FlowerHeadRadius);
            Assert.Equal(a[i].HeightBonus, b[i].HeightBonus);
        }
    }

    [Fact]
    public void FlowerCountIsWithin3SigmaOfProbability()
    {
        var blades = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, 1.0);
        Assert.True(blades.Length > 100);
        int flowerCount = blades.Count(b => b.IsFlower);
        double n  = blades.Length;
        double p  = Constants.FLOWER_PROBABILITY;
        double mu = n * p;
        double sd = Math.Sqrt(n * p * (1.0 - p));
        Assert.InRange(flowerCount, (int)Math.Floor(mu - 3.0 * sd), (int)Math.Ceiling(mu + 3.0 * sd));
    }

    [Fact]
    public void FlowerStreamDoesNotPerturbMainStream()
    {
        var blades = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, 1.0);
        Assert.True(blades.Length > 0);
        Assert.Equal(4.941073726820111, blades[0].BaseX, 12);
        Assert.Equal(24.469991818248864, blades[0].Height, 12);
        Assert.Equal(1.5829214329729786, blades[0].Thickness, 12);
        Assert.Equal((byte)3, blades[0].Hue);
    }

    [Fact]
    public void NonFlowerBladesHaveUnitHeightBonus()
    {
        var blades = Sim.GenerateBlades(Constants.CANONICAL_TEST_SEED, Monitor1920, 1.0);
        foreach (ref readonly var b in blades.AsSpan())
        {
            if (!b.IsFlower)
            {
                Assert.Equal(1.0, b.HeightBonus);
                Assert.Equal(0.0, b.FlowerHeadRadius);
            }
        }
    }
}

File: tests/DesktopGrass.Native.Tests/src/flower_tests.cpp
// flower_tests.cpp
//
// Tests for §5 flower stream + §7 head-render contract.

#include "TestHelpers.h"
#include "Sim.h"

#include <cmath>
#include <vector>

using namespace desktopgrass;

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(FlowerTests)
{
public:
TEST_METHOD(FlowerStreamIsDeterministicForAGivenSeed) {
    std::vector<Blade> a, b;
    generate_blades(CANONICAL_TEST_SEED, 1920.0, 1.0, a);
    generate_blades(CANONICAL_TEST_SEED, 1920.0, 1.0, b);
    Assert::IsTrue(a.size() == b.size());
    for (std::size_t i = 0; i < a.size(); ++i) {
        Assert::IsTrue(a[i].isFlower           == b[i].isFlower);
        Assert::IsTrue(a[i].flowerHeadColorIdx == b[i].flowerHeadColorIdx);
        Assert::IsTrue(a[i].flowerHeadRadius   == b[i].flowerHeadRadius);
        Assert::IsTrue(a[i].heightBonus        == b[i].heightBonus);
    }
}

TEST_METHOD(FlowerCountIsWithin3SigmaOfFLOWERPROBABILITY) {
    std::vector<Blade> blades;
    generate_blades(CANONICAL_TEST_SEED, 1920.0, 1.0, blades);
    Assert::IsTrue(blades.size() > 100);

    std::size_t flowerCount = 0;
    for (const Blade& b : blades) if (b.isFlower) ++flowerCount;

    const double n   = static_cast<double>(blades.size());
    const double p   = FLOWER_PROBABILITY;
    const double mu  = n * p;
    const double sd  = std::sqrt(n * p * (1.0 - p));
    // 3-sigma tolerance keeps this test stable across spec-conformant
    // PRNG sequences. For seed=0x6B6173746F, n=321 we expect ~12.84 with
    // sd≈3.51, so [2,24] is the acceptable range.
    Assert::IsTrue(flowerCount >= static_cast<std::size_t>(std::floor(mu - 3.0 * sd)));
    Assert::IsTrue(flowerCount <= static_cast<std::size_t>(std::ceil(mu + 3.0 * sd)));
}

TEST_METHOD(FlowerStreamDoesNotPerturbTheMainStream) {
    // Regenerate blades and assert the main-stream fields match the
    // canonical snapshot. This is implicitly covered by blade_gen_tests
    // (the first/last 10 still match), but pin it here for clarity.
    std::vector<Blade> blades;
    generate_blades(CANONICAL_TEST_SEED, 1920.0, 1.0, blades);
    Assert::IsTrue(blades.size() > 0);
    Assert::IsTrue(blades[0].baseX == Near(4.941073726820111).margin(1e-12));
    Assert::IsTrue(blades[0].height == Near(24.469991818248864).margin(1e-12));
}
};
}
